### PR TITLE
Added DropValidator task.

### PR DESCRIPTION
**Pull request by aasim**
**Task name**: DropValidator

**Description**: drop validator initial check in

**Documentation changes required:** (Y/N) N

**Added unit tests:** (Y/N) n, just calls the tool which has unit tests

**Attached related issue:** (Y/N) 

**Checklist**:
- [X] Task version was bumped - please check [instruction](https://github.com/microsoft/azure-pipelines-tasks/tree/master/docs/taskversionbumping.md) how to do it
- [X] Checked that applied changes work as expected

### CODE REVIEW

**Comment by aasim** [comment]
https://dev.azure.com/mseng/_git/AzureDevOps/pullrequest/592967?_a=files will publish this once I push out the nuget package

File: Tasks/DropValidatorV0/Strings/resources.resjson/en-US/resources.resjson [hunk 1]
```diff
@@ -0,0 +1,17 @@
+{
+  "loc.friendlyName": "Drop validator",
+  "loc.helpMarkDown": "",
+  "loc.description": "Validates a given manifest with a manifest file to verify its authenticity.",
+  "loc.instanceNameFormat": "Drop Validator $(definition)",
```
Comment (edgarrs) [inline Tasks/DropValidatorV0/Strings/resources.resjson/en-US/resources.resjson:5]: In case we want to rename the project, I think we should use a variable to store the name so we can just replace the value of variable and have everything renamed. Also, this will help with inconsistent casing of the name: Drop Validator, drop validator, DropValidator, etc.

File: Tasks/DropValidatorV0/main.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+async function getDropValidator() {
+    try {
+        // Get dotnet tool
+        const dotnetPath = tl.which('dotnet', true);
+        const dotnet = tl.tool(dotnetPath);
+
+        // Install or update the dropvalidator tool globally.
+        dotnet.line('tool update Microsoft.DropValidator --global');
```
Comment (edgarrs) [inline Tasks/DropValidatorV0/main.ts:37]: put this in a variable

File: Tasks/DropValidatorV0/main.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
+import * as tl from 'azure-pipelines-task-lib/task';
+import * as telemetry from 'azure-pipelines-tasks-utility-common/telemetry';
+import * as fs from 'fs';
+
+async function run() {
+    try {
+        const dropValidatorArgs = ['BuildDropPath', 'ManifestPath', 'OutputPath', 'Verbosity', 'ConfigFilePath', 'RootPathFilter'];
```
Comment (edgarrs) [inline Tasks/DropValidatorV0/main.ts:7]: Do we validate the OutputPath inside the dropvalidator code? I'm just curious if everything will succeed except trying to write the results because the OutputPath is incorrect.

File: Tasks/DropValidatorV0/package.json [hunk 4]
```diff
@@ -0,0 +1,26 @@
+{
+  "name": "dropvalidator",
+  "version": "0.1.0",
+  "description": "Validates a given manifest with a manifest file to verify its authenticity.",
```
Comment (edgarrs) [inline Tasks/DropValidatorV0/package.json:4]: can you please reword this to make it clearer?

File: Tasks/DropValidatorV0/task.json [hunk 5]
```diff
@@ -0,0 +1,92 @@
+{
+    "$schema": "https://raw.githubusercontent.com/Microsoft/azure-pipelines-task-lib/master/tasks.schema.json",
+    "id": "bae07fbb-2096-488a-ab37-adf7808cc9fe",
+    "name": "DropValidator",
+    "friendlyName": "Drop validator",
+    "description": "Validates a given manifest with a manifest file to verify its authenticity.",
```
Comment (edgarrs) [inline Tasks/DropValidatorV0/task.json:6]: This looks like a recurrent message in different files. Is it possible to store it in a variable?

**Review by edgarrs — CHANGES_REQUESTED** [review]

File: Tasks/DropValidatorV0/main.ts (hunk 2 above)
Comment (jmyersmsft) [inline Tasks/DropValidatorV0/main.ts:37]: modifying global machine state in a task is problematic. For one thing, multiple agents running on the same machine is a supported scenario.

Comment (jmyersmsft) [reply Tasks/DropValidatorV0/main.ts:37]: Downloading tools from the internet is also problematic, because running TFS in a no-internet environment is also a supported scenario.

File: Tasks/DropValidatorV0/main.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        dotnet.line('tool update Microsoft.DropValidator --global');
+        const result = await dotnet.exec();
+        tl.debug('Result: ' + result);
+
+        if (result === 0) {
+            const dropValidatorPath = tl.which('dropvalidator', true);
+            return tl.tool(dropValidatorPath);
+        }
+
+        throw new Error();
```
Comment (jmyersmsft) [inline Tasks/DropValidatorV0/main.ts:46]: The Error object should probably be given a message

File: Tasks/DropValidatorV0/package.json [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+  "main": "main.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/Microsoft/vsts-tasks.git"
```
Comment (jmyersmsft) [inline Tasks/DropValidatorV0/package.json:11]: I don't think this is the canonical address of the repo anymore

File: Tasks/DropValidatorV0/task.json [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    "helpMarkDown": "",
+    "category": "Utility",
+    "author": "ms-vscs-rm",
```
Comment (jmyersmsft) [inline Tasks/DropValidatorV0/task.json:9]: this author value looks suspicious

File: Tasks/DropValidatorV0/task.json (hunk 5 above)
Comment (jmyersmsft) [inline Tasks/DropValidatorV0/task.json:6]: "manifest" twice here is confusing

Comment (jmyersmsft) [reply Tasks/DropValidatorV0/task.json:6]: There are no variables in task.json files

Comment (edgarrs) [reply Tasks/DropValidatorV0/task.json:6]: I see $(definition) below in the file, isn't that a kind of variable? I confess that I have never written a task and I'm not sure how this task.json is used and if there is a process to do some replace of some values or something like that

Comment (jmyersmsft) [reply Tasks/DropValidatorV0/task.json:6]: that's specific to that particular property. It's normally used in the editor to substitute task parameters so the name of the task magically includes one or more of the parameter values. I'm not sure what "definition" would expand to though...

File: Tasks/DropValidatorV0/main.ts (hunk 3 above)
Comment (aasim) [reply Tasks/DropValidatorV0/main.ts:7]: The value is a relative path from the build agent run folder, and it's writable. We also check on the dropvalidator.exe to see if this location is not writable and log an error accordingly.

File: Tasks/DropValidatorV0/main.ts (hunk 2 above)
Comment (aasim) [reply Tasks/DropValidatorV0/main.ts:37]: I can definitely try to make the tool install local, but if the ask is to not download anything from the internet, this might a be a problem in the future. The only other way would be to distribute the nupkg or the exe with the task, like how we distribute the 7zip exe. Is there any other way you know that we can implement this?

File: Tasks/DropValidatorV0/Strings/resources.resjson/en-US/resources.resjson (hunk 1 above)
Comment (aasim) [reply Tasks/DropValidatorV0/Strings/resources.resjson/en-US/resources.resjson:5]: The value can't be a variable, we will have to change all occurrences if we change the name in the future. We will also have to create a new task if the name has to be changed from DropValidator to something else.

File: Tasks/DropValidatorV0/Strings/resources.resjson/en-US/resources.resjson [hunk 9]
```diff
@@ -0,0 +1,17 @@
+{
+  "loc.friendlyName": "Drop validator",
+  "loc.helpMarkDown": "",
+  "loc.description": "Validates a given manifest with a manifest file to verify its authenticity.",
```
[the remark this answers is not in the document]
Comment (aasim) [reply Tasks/DropValidatorV0/Strings/resources.resjson/en-US/resources.resjson:4]: Fixed

File: Tasks/DropValidatorV0/main.ts (hunk 2 above)
Comment (aasim) [reply Tasks/DropValidatorV0/main.ts:37]: fixed

File: Tasks/DropValidatorV0/task.json (hunk 5 above)
Comment (aasim) [reply Tasks/DropValidatorV0/task.json:6]: fixed

File: Tasks/DropValidatorV0/package.json (hunk 4 above)
Comment (aasim) [reply Tasks/DropValidatorV0/package.json:4]: fixed

File: Tasks/DropValidatorV0/package.json (hunk 7 above)
Comment (aasim) [reply Tasks/DropValidatorV0/package.json:11]: Fixed

File: Tasks/DropValidatorV0/task.json (hunk 5 above)
Comment (aasim) [reply Tasks/DropValidatorV0/task.json:6]: fixed
